citation_parser.py

```diff
@@ -0,0 +1,130 @@
+import re
+from collections import namedtuple
+from types import SimpleNamespace
+from nltk.tokenize import sent_tokenize
+
+Citation = namedtuple('Citation', ['text', 'prev', 'start', 'end'])
+
+class CitationParser(object):
+
+    def __init__(self):
+        self._ref_regex = '<ref.+?(/>|</ref>)'
+        self._cn_regex = '{{(C|c)itation needed\|.+?}}'
+        self._citation_regex = self._ref_regex + '|' + self._cn_regex
+        self._citations = []
+        self._sentences = []
+
+    def _find_citations(self, wikitext):
+        for ref in re.finditer(self._citation_regex, wikitext):
+            self._citations.append(Citation(text=ref.group(0), prev=wikitext[ref.start(0)-1], 
+                start=ref.start(0), end=ref.start(0)+len(ref.group(0))))
+        self._citations.reverse()
+
+    def _remove_citation_tags(self, wikitext):
+        return re.sub(self._citation_regex, '', wikitext)
+
+    def _add_successive_citation(self):
+        while self._citations:
+            if self._citations[-1].prev == '>':
+                self._sentences[-1].cite.append(self._citations.pop())
+            else:
+                break
+
+    def parse(self, wikitext):
+        """
+        This is a method for parsing wikitext to get individual sentences
+        and the citations they have.
+        
+        The algorithm goes like:
+        1) Find reference tags <ref> and citation needed templates {cn}
+           within the wikitext and store them in a queue in context order.
+        2) Get the plain text by removing <ref> and {cn}.
+        3) Split the plain text into sentences.
+        4) Loop over the sentences to see whether we can find the sentence
+           in the wikitext.
+           -If found, that means the sentence isn't broken by <ref> or
+            {cn}, then we check whether any citations is followed behind
+            the sentence.
+           -If not found, the sentence is broken by <ref> or {cn}, we find
+            the correct start location and end location of the sentence
+            and also check whether any citation is followed behind.
```

Comment:
So as we discussed today, a `{{cn}}` or a `<ref>` don't necessarily apply just to the sentence right before it. They could apply to more than one sentence, or a whole paragraph (or maybe more?) and there's no way to determine that from the Wikitext.

Given how hard it is to get this detection correctly, I wonder if we really really want this optimization? What will happen if we don't merge this PR? Does the job get intolerably slow? Or does this severely reduce the quality of the data we produce?

Wikitext is, unfortunately, a difficult beast to tame, and I'd like to avoid spending a lot of our time doing fancy processing of it unless we really have to. The fact that CH actually uses HTML is a consequence of that: I spent a lot of time trying to clean up Wikitext in various ways like this, until I eventually gave up and realized I'd only get the data quality I wanted if I just switched to HTML. For CD, it would be nice to keep things simple for now if we can -- if this is too slow, we can always tweak the input to do more aggressive sampling, or only choose articles within a certain category etc.

Reply:
> So as we discussed today, a `{{cn}}` or a `<ref>` don't necessarily apply just to the sentence right before it. They could apply to more than one sentence, or a whole paragraph (or maybe more?) and there's no way to determine that from the Wikitext.
>
Yeah indeed, that's what I didn't think of. 
I was thinking every input for Citation Needed model is "one" sentence, so that's why I tried to split a paragraph into sentences and determine whether the sentence has a reference or not.

But when I check the training data, I found they are not necessarily one sentence, sometimes are two sentences or even a whole paragraph and I'm not really sure how they determine that from HTML.  Now I worry it is appropriate to use `sent_tokenize` to split a paragraph into individual sentences.

If we don't merge this PR, then we will run all the individual sentences in an article no matter there is a reference or not, but it seems not a big deal because the number of referenced sentences is just a small portion of an article in most cases, and actually this PR  runs slower because of more text processing steps.

Reply:
Oh, interesting, I didn't know about the training data either! I think we can keep using sentences for now though.

And yeah, I think running all sentences through the model is a good start. So I guess we're not going ahead with this PR, then?